File: Sources/cmd-m/LayoutConverter.swift
import Carbon
import Foundation

/// A physical key press: hardware key code + whether Shift is held.
struct KeyStroke: Hashable {
    let keyCode: UInt16
    let shift: Bool
}

/// One enabled keyboard layout, with maps between key strokes and the
/// characters they produce, derived from the system's own layout data.
final class Layout {
    let inputSource: TISInputSource
    let inputSourceID: String
    let localizedName: String
    private(set) var charForStroke: [KeyStroke: Character] = [:]
    private(set) var strokeForChar: [Character: KeyStroke] = [:]

    init?(source: TISInputSource) {
        inputSource = source
        guard
            let id: String = Layout.property(source, kTISPropertyInputSourceID),
            let name: String = Layout.property(source, kTISPropertyLocalizedName),
            let layoutData: Data = Layout.property(source, kTISPropertyUnicodeKeyLayoutData)
        else { return nil }

        inputSourceID = id
        localizedName = name

        let kbdType = UInt32(LMGetKbdType())
        layoutData.withUnsafeBytes { (raw: UnsafeRawBufferPointer) in
            guard let layoutPtr = raw.baseAddress?.assumingMemoryBound(to: UCKeyboardLayout.self) else { return }
            for keyCode: UInt16 in 0..<128 {
                for shift in [false, true] {
                    // UCKeyTranslate expects (EventModifiers >> 8) & 0xFF; shiftKey is 0x200.
                    let modifiers: UInt32 = shift ? UInt32((shiftKey >> 8) & 0xFF) : 0
                    var deadKeyState: UInt32 = 0
                    var length = 0
                    var chars = [UniChar](repeating: 0, count: 4)
                    let status = UCKeyTranslate(
                        layoutPtr,
                        keyCode,
                        UInt16(kUCKeyActionDown),
                        modifiers,
                        kbdType,
                        UInt32(1 << kUCKeyTranslateNoDeadKeysBit),
                        &deadKeyState,
                        chars.count,
                        &length,
                        &chars
                    )
                    guard status == noErr, length == 1,
                          let scalar = UnicodeScalar(chars[0]),
                          scalar.value >= 0x20, scalar.value != 0x7F
                    else { continue }

                    let ch = Character(scalar)
                    let stroke = KeyStroke(keyCode: keyCode, shift: shift)
                    charForStroke[stroke] = ch
                    if strokeForChar[ch] == nil {
                        strokeForChar[ch] = stroke
                    }
                }
            }
        }

        if strokeForChar.isEmpty { return nil }
    }

    private static func property<T>(_ source: TISInputSource, _ key: CFString!) -> T? {
        guard let ptr = TISGetInputSourceProperty(source, key) else { return nil }
        let value = Unmanaged<AnyObject>.fromOpaque(ptr).takeUnretainedValue()
        if T.self == Data.self {
            return (value as? Data) as? T
        }
        return value as? T
    }
}

enum LayoutConverter {

    /// All keyboard layouts the user has enabled in System Settings, in the
    /// system's ordering. Input methods (e.g. Chinese, Japanese) have no
    /// static key map and are skipped.
    static func enabledLayouts() -> [Layout] {
        let filter = [kTISPropertyInputSourceType: kTISTypeKeyboardLayout] as CFDictionary
        guard let cfList = TISCreateInputSourceList(filter, false)?.takeRetainedValue() else {
            return []
        }
        let sources = cfList as! [TISInputSource]
        var layouts: [Layout] = []
        for source in sources {
            if let layout = Layout(source: source) {
                // Skip duplicates that map to an already-seen input source.
                if !layouts.contains(where: { $0.inputSourceID == layout.inputSourceID }) {
                    layouts.append(layout)
                }
            }
        }
        return layouts
    }

    struct Conversion {
        let text: String
        let target: Layout
    }

    /// Re-types `text` from the layout it appears to have been written in
    /// into the next enabled layout, cycling through the user's layouts.
    /// Returns nil when fewer than two layouts are enabled.
    static func convert(_ text: String) -> Conversion? {
        let layouts = enabledLayouts()
        guard layouts.count >= 2 else { return nil }

        let sourceIndex = bestSourceLayout(for: text, in: layouts)
        let source = layouts[sourceIndex]
        let target = layouts[(sourceIndex + 1) % layouts.count]

        var result = ""
        result.reserveCapacity(text.count)
        for ch in text {
            if let stroke = source.strokeForChar[ch],
               let mapped = target.charForStroke[stroke] {
                result.append(mapped)
            } else {
                result.append(ch)
            }
        }
        return Conversion(text: result, target: target)
    }

    /// Picks the layout whose key map covers the most characters of `text` —
    /// i.e. the layout the text was most plausibly typed in.
    private static func bestSourceLayout(for text: String, in layouts: [Layout]) -> Int {
        var bestIndex = 0
        var bestScore = -1
        for (index, layout) in layouts.enumerated() {
            var score = 0
            for ch in text where layout.strokeForChar[ch] != nil {
                score += 1
            }
            if score > bestScore {
                bestScore = score
                bestIndex = index
            }
        }
        return bestIndex
    }
}

File: windows/LayoutConverter.cs
using System.Globalization;
using System.Runtime.InteropServices;
using System.Text;

namespace Relayout;

/// A physical key press: hardware scan code + whether Shift is held.
/// Scan codes (not virtual keys) make conversion positional, matching the
/// macOS implementation — so QWERTY↔AZERTY etc. convert by key position.
readonly record struct KeyStroke(uint Scan, bool Shift);

/// One installed keyboard layout, with maps between key strokes and the
/// characters they produce, derived from the system's own layout data.
sealed class KeyboardLayout
{
    public IntPtr Hkl { get; }
    public string Name { get; }
    public Dictionary<KeyStroke, char> CharForStroke { get; } = new();
    public Dictionary<char, KeyStroke> StrokeForChar { get; } = new();

    // The typing-area scan codes: digit row, three letter rows with their
    // punctuation neighbors, backslash, backtick, and space.
    static readonly uint[] MappedScanCodes = BuildScanCodes();

    static uint[] BuildScanCodes()
    {
        var scans = new List<uint> { 0x29, 0x2B, 0x39 }; // ` \ space
        for (uint sc = 0x02; sc <= 0x0D; sc++) scans.Add(sc); // 1..0 - =
        for (uint sc = 0x10; sc <= 0x1B; sc++) scans.Add(sc); // q..p [ ]
        for (uint sc = 0x1E; sc <= 0x28; sc++) scans.Add(sc); // a..l ; '
        for (uint sc = 0x2C; sc <= 0x35; sc++) scans.Add(sc); // z..m , . /
        return scans.ToArray();
    }

    public KeyboardLayout(IntPtr hkl)
    {
        Hkl = hkl;
        Name = LayoutName(hkl);

        var state = new byte[256];
        var buffer = new StringBuilder(8);
        foreach (var scan in MappedScanCodes)
        {
            uint vk = Native.MapVirtualKeyEx(scan, Native.MAPVK_VSC_TO_VK_EX, hkl);
            if (vk == 0) continue;
            foreach (var shift in new[] { false, true })
            {
                Array.Clear(state);
                if (shift) state[0x10] = 0x80; // VK_SHIFT
                buffer.Clear();
                // 0x4: don't change kernel keyboard state (Win10 1607+), so
                // probing dead keys can't corrupt the user's typing.
                int rc = Native.ToUnicodeEx(vk, scan, state, buffer, buffer.Capacity, 0x4, hkl);
                if (rc != 1) continue; // dead key (<0), nothing (0), or multi-char
                char ch = buffer[0];
                if (ch < 0x20 || ch == 0x7F) continue;

                var stroke = new KeyStroke(scan, shift);
                CharForStroke[stroke] = ch;
                if (!StrokeForChar.ContainsKey(ch)) StrokeForChar[ch] = stroke;
            }
        }
    }

    static string LayoutName(IntPtr hkl)
    {
        try
        {
            int langId = (int)hkl & 0xFFFF;
            return CultureInfo.GetCultureInfo(langId).DisplayName;
        }
        catch
        {
            return $"0x{(long)hkl:X}";
        }
    }
}

sealed record Conversion(string Text, IntPtr TargetHkl, string TargetName);

static class LayoutConverter
{
    /// All keyboard layouts installed for this user, in system order.
    public static List<KeyboardLayout> InstalledLayouts()
    {
        int count = Native.GetKeyboardLayoutList(0, null);
        if (count <= 0) return new List<KeyboardLayout>();
        var hkls = new IntPtr[count];
        Native.GetKeyboardLayoutList(count, hkls);
        var layouts = new List<KeyboardLayout>();
        foreach (var hkl in hkls)
        {
            var layout = new KeyboardLayout(hkl);
            if (layout.StrokeForChar.Count > 0) layouts.Add(layout);
        }
        return layouts;
    }

    /// Re-types `text` from the layout it appears to have been written in
    /// into the next installed layout, cycling. Null when fewer than two
    /// layouts are installed.
    public static Conversion? Convert(string text) => ConvertWith(text, InstalledLayouts());

    /// Same conversion against an explicit layout list — lets tests exercise
    /// language pairs in isolation, regardless of what the machine has loaded.
    public static Conversion? ConvertWith(string text, List<KeyboardLayout> layouts)
    {
        if (layouts.Count < 2) return null;

        int sourceIndex = BestSourceLayout(text, layouts);
        var source = layouts[sourceIndex];
        var target = layouts[(sourceIndex + 1) % layouts.Count];

        var result = new StringBuilder(text.Length);
        foreach (char ch in text)
        {
            if (source.StrokeForChar.TryGetValue(ch, out var stroke)
                && target.CharForStroke.TryGetValue(stroke, out var mapped))
            {
                result.Append(mapped);
            }
            else
            {
                result.Append(ch);
            }
        }
        return new Conversion(result.ToString(), target.Hkl, target.Name);
    }

    /// The layout whose key map covers the most characters of `text` —
    /// i.e. the layout the text was most plausibly typed in.
    static int BestSourceLayout(string text, List<KeyboardLayout> layouts)
    {
        int bestIndex = 0, bestScore = -1;
        for (int i = 0; i < layouts.Count; i++)
        {
            int score = 0;
            foreach (char ch in text)
                if (layouts[i].StrokeForChar.ContainsKey(ch)) score++;
            if (score > bestScore) { bestScore = score; bestIndex = i; }
        }
        return bestIndex;
    }
}
